MantidQt/MantidWidgets/src/MantidWSIndexDialog.cpp

```diff
@@ -243,21 +410,152 @@ void MantidWSIndexWidget::initSpectraBox() {
     if (m_waterfall) {
-      m_plotOptions->addItem(tr("Waterfall Plot"));
+      m_plotOptions->addItem(WATERFALL_PLOT);
     }
     if (m_tiled) {
       m_plotOptions->addItem(tr("Tiled Plot"));
     }
+    if (m_advanced && isSuitableForContourOrSurfacePlot()) {
+      m_plotOptions->addItem(SURFACE_PLOT);
+      m_plotOptions->addItem(CONTOUR_PLOT);
+      connect(m_plotOptions, SIGNAL(currentIndexChanged(const QString &)), this,
+              SLOT(onPlotOptionChanged(const QString &)));
+    }
     m_optionsBox->addWidget(m_plotOptions);
   }
 
   m_outer->addItem(m_optionsBox);
 }
 
+void MantidWSIndexWidget::initLogs() {
+  m_logBox = new QVBoxLayout;
+
+  m_logLabel = new QLabel(tr("Log value to plot against:"));
+  m_logSelector = new QComboBox();
+  populateLogComboBox();
+
+  m_customLogLabel = new QLabel(tr("<br>Custom log values:"));
+  m_logValues = new QLineEdit();
+
+  m_axisLabel = new QLabel(tr("<br>Label for plot axis:"));
+  m_axisNameEdit = new QLineEdit();
+  m_axisNameEdit->setText(m_logSelector->currentText());
+
+  m_logBox->addWidget(m_logLabel);
+  m_logBox->addWidget(m_logSelector);
+  m_logBox->addWidget(m_customLogLabel);
+  m_logBox->addWidget(m_logValues);
+  m_logBox->addWidget(m_axisLabel);
+  m_logBox->addWidget(m_axisNameEdit);
+
+  m_logSelector->setEnabled(true);
+  m_logValues->setEnabled(true);
+  m_axisNameEdit->setEnabled(false);
+
+  m_outer->addItem(m_logBox);
+
+  connect(m_logSelector, SIGNAL(currentIndexChanged(const QString &)), this,
+          SLOT(onLogSelected(const QString &)));
+}
+
+/**
+* Called when log selection changed
+* If "Custom" selected, enable the custom log input box.
+* Otherwise, it is read-only.
+* Also put the log name into the axis name box as a default choice.
+* @param logName :: [input] Text selected in combo box
+*/
+void MantidWSIndexWidget::onLogSelected(const QString &logName) {
+  m_logValues->setEnabled(logName == CUSTOM);
+  m_logValues->clear();
+  m_axisNameEdit->setText(logName);
+}
+
+/**
+* Called when plot option is changed
+* @param plotOption :: [input] New plot option
+*/
+void MantidWSIndexWidget::onPlotOptionChanged(const QString &plotOption) {
+  auto useLogNames = m_advanced && isSuitableForLogValues(plotOption);
+  auto isLogSelectorCustom = m_logSelector->currentText() == CUSTOM;
+  m_logSelector->setEnabled(useLogNames);
+  m_logValues->setEnabled(useLogNames && isLogSelectorCustom);
+  m_logValues->clear();
+  m_axisNameEdit->setEnabled(useLogNames);
+}
+
+/**
+* Populate the log combo box with all log names that
+* have single numeric value per workspace (and occur
+* in every workspace)
+*/
+void MantidWSIndexWidget::populateLogComboBox() {
+  // First item should be "Workspace index"
+  m_logSelector->addItem(WORKSPACE_INDEX);
+
+  // Create a table of all single-value numeric log names versus
+  // how many workspaces they appear in
+  std::map<std::string, int> logCounts;
+  for (auto wsName : m_wsNames) {
```

Comment:
take this by reference, ie `auto& wsName : m_wsNames`. This avoids the copy of each element